File: app/src/main/java/jp/juggler/fadownloader/LogViewer.java
package jp.juggler.fadownloader;

import android.content.Context;
import android.database.Cursor;
import android.os.Bundle;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CursorAdapter;
import android.widget.ListView;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.TimeZone;
import java.util.GregorianCalendar;

public class LogViewer implements LoaderManager.LoaderCallbacks<Cursor>{

	AppCompatActivity activity;
	ListView listview;
	int last_view_start;
	Loader<Cursor> loader;

	public LogViewer(){
	}

	void onStart( AppCompatActivity activity, ListView listview, int loader_id ){
		this.activity = activity;
		this.listview = listview;
		loader = activity.getSupportLoaderManager().initLoader( loader_id, null, this );
	}

	void onStop(){
		if( loader != null ){
			activity.getSupportLoaderManager().destroyLoader( loader.getId() );
			loader = null;
		}
		this.last_view_start = listview.getFirstVisiblePosition();
		listview.setAdapter( null );
		this.listview = null;
		this.activity = null;
	}

	@Override public Loader<Cursor> onCreateLoader( int id, Bundle args ){
		return activity == null ? null : new CursorLoader( activity, LogData.meta.content_uri, null, null, null, LogData.COL_TIME + " desc" );
	}

	@Override public void onLoadFinished( Loader<Cursor> loader, Cursor cursor ){
		if( listview == null ) return;
		LogAdapter adapter = (LogAdapter) listview.getAdapter();
		if( adapter == null ){
			listview.setAdapter( new LogAdapter( activity, cursor ) );
		}else{
			adapter.swapCursor( cursor );
		}
	}

	@Override public void onLoaderReset( Loader<Cursor> loader ){
		if( listview == null ) return;
		LogAdapter adapter = (LogAdapter) listview.getAdapter();
		if( adapter != null ){
			adapter.swapCursor( null );
		}
	}

	static class ViewHolder{

		TextView tvTime;
		TextView tvMessage;
	}

	class LogAdapter extends CursorAdapter{

		LayoutInflater inflater = activity.getLayoutInflater();
		int colidx_time;
		int colidx_message;
		int colidx_level;

		SimpleDateFormat date_fmt = new SimpleDateFormat( "yyyy-MM-dd HH:mm:ss.SSS z", Locale.getDefault(  ) );

		public LogAdapter( Context context, Cursor c ){
			super( context, c, false );
			//
			colidx_time = c.getColumnIndex( LogData.COL_TIME );
			colidx_message = c.getColumnIndex( LogData.COL_MESSAGE );
			colidx_level = c.getColumnIndex( LogData.COL_LEVEL );

		}

		@Override public View newView( Context context, Cursor cursor, ViewGroup viewGroup ){
			View root = inflater.inflate( R.layout.lv_log, viewGroup, false );
			ViewHolder holder = new ViewHolder();
			holder.tvTime = (TextView) root.findViewById( R.id.tvTime );
			holder.tvMessage = (TextView) root.findViewById( R.id.tvMessage );
			root.setTag( holder );
			return root;
		}

		@Override public void bindView( View view, Context context, Cursor cursor ){
			ViewHolder holder = (ViewHolder) view.getTag();
			long time = cursor.getLong( colidx_time );
			String message = cursor.getString( colidx_message );
			int level = cursor.getInt( colidx_level );
			holder.tvTime.setText( date_fmt.format(time ) );
			holder.tvMessage.setText( message );

			int fg;
			if( level >= LogData.LEVEL_FLOOD ){
				fg = 0xff777777;
			}else if( level >= LogData.LEVEL_HEARTBEAT ){
				fg = 0xff555555;
			}else if( level >= LogData.LEVEL_DEBUG ){
				fg = 0xff333333;
			}else if( level >= LogData.LEVEL_VERBOSE ){
				fg = 0xff111111;
			}else if( level >= LogData.LEVEL_INFO ){
				fg = 0xff000000;
			}else if( level >= LogData.LEVEL_WARNING ){
				fg = 0xffff8000;
			}else{
				fg = 0xffff0000;
			}
			holder.tvTime.setTextColor( fg );
			holder.tvMessage.setTextColor( fg );
		}
	}

}

File: app/src/main/java/jp/juggler/fadownloader/LogViewer.kt
package jp.juggler.fadownloader

import android.content.Context
import android.database.Cursor
import android.os.Bundle
import android.support.v4.app.LoaderManager
import android.support.v4.content.CursorLoader
import android.support.v4.content.Loader
import android.support.v7.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CursorAdapter
import android.widget.ListView
import android.widget.TextView
import jp.juggler.fadownloader.table.LogData

import java.text.SimpleDateFormat
import java.util.Locale

class LogViewer {
	
	companion object {
		
		internal val date_fmt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS z", Locale.getDefault())
	}
	
	class Holder(
		val activity : AppCompatActivity,
		private val listView : ListView
	) : LoaderManager.LoaderCallbacks<Cursor> {
		
		private var loader : Loader<Cursor>? = null
		
		fun onStart(loader_id : Int) {
			loader = activity.supportLoaderManager.initLoader(loader_id, null, this)
		}
		
		fun onStop() : Int {
			val loader = this.loader
			if(loader != null) {
				activity.supportLoaderManager.destroyLoader(loader.id)
				this.loader = null
			}
			val rv = listView.firstVisiblePosition
			listView.adapter = null
			return rv
		}
		
		override fun onCreateLoader(id : Int, args : Bundle?) : Loader<Cursor> {
			return CursorLoader(
				activity,
				LogData.meta.content_uri,
				null,
				null,
				null,
				LogData.COL_TIME + " desc"
			)
		}
		
		override fun onLoadFinished(loader : Loader<Cursor>, cursor : Cursor) {
			val adapter = listView.adapter as? LogAdapter
			if(adapter == null) {
				listView.adapter = LogAdapter(cursor)
			} else {
				adapter.swapCursor(cursor)
			}
		}
		
		override fun onLoaderReset(loader : Loader<Cursor>) {
			val adapter = listView.adapter as? LogAdapter
			adapter?.swapCursor(null)
		}
		
		internal inner class LogAdapter(c : Cursor) : CursorAdapter(activity, c, false) {
			
			private val inflater : LayoutInflater = activity.layoutInflater
			private val colidx_time : Int = c.getColumnIndex(LogData.COL_TIME)
			private val colidx_message : Int = c.getColumnIndex(LogData.COL_MESSAGE)
			private val colidx_level : Int = c.getColumnIndex(LogData.COL_LEVEL)
			
			override fun newView(context : Context, cursor : Cursor, viewGroup : ViewGroup) : View {
				val root = inflater.inflate(R.layout.lv_log, viewGroup, false)
				val holder = ViewHolder(root)
				root.tag = holder
				return root
			}
			
			override fun bindView(view : View, context : Context, cursor : Cursor) {
				val holder = (view.tag as ViewHolder)
				
				val time = cursor.getLong(colidx_time)
				val message = cursor.getString(colidx_message)
				
				val level = cursor.getInt(colidx_level)
				val fg : Int
				fg = when {
					level >= LogData.LEVEL_FLOOD -> - 0x444445
					level >= LogData.LEVEL_HEARTBEAT -> - 0x666667
					level >= LogData.LEVEL_DEBUG -> - 0x888889
					level >= LogData.LEVEL_VERBOSE -> - 0xaaaaab
					level >= LogData.LEVEL_INFO -> - 0x1000000
					level >= LogData.LEVEL_WARNING -> - 0x8000
					else -> - 0x10000
				}
				
				holder.tvTime.text = date_fmt.format(time)
				holder.tvMessage.text = message
				holder.tvTime.setTextColor(fg)
				holder.tvMessage.setTextColor(fg)
			}
		}
		
		internal class ViewHolder(root : View) {
			val tvTime : TextView = root.findViewById(R.id.tvTime)
			val tvMessage : TextView = root.findViewById(R.id.tvMessage)
		}
	}
	
	private var last_view_start : Int = 0
	
	private var holder : Holder? = null
	
	internal fun onStart(activity : AppCompatActivity, listView : ListView, loader_id : Int) {
		val holder = Holder(activity, listView)
		this.holder = holder
		holder.onStart(loader_id)
	}
	
	internal fun onStop() {
		val rv = holder?.onStop()
		holder = null
		if(rv != null) last_view_start = rv
	}
	
}
